asdf_astropy/converters/transform/polynomial.py

```diff
@@ -0,0 +1,187 @@
+from packaging.version import parse as parse_version
+import numpy as np
+
+from .core import TransformConverterBase, parameter_to_value
+
+from ..helpers import parse_tag_version
+
+
+class PolynomialConverter(TransformConverterBase):
+    """
+    ASDF support for serializing the 1D and 2D polynomial models.
+    """
+    # Schema versions prior to 1.2 included an unrelated "domain"
+    # property.  We can't serialize the new domain values with those
+    # versions because they don't validate.
+    _DOMAIN_WINDOW_MIN_VERSION = parse_version("1.2.0")
+
+    tags = ["tag:stsci.edu:asdf/transform/polynomial-*"]
+
+    types = [
+        "astropy.modeling.polynomial.Polynomial1D",
+        "astropy.modeling.polynomial.Polynomial2D",
+    ]
+
+    def to_yaml_tree_transform(self, model, tag, ctx):
+        from astropy.modeling.polynomial import Polynomial1D, Polynomial2D
+
+        if isinstance(model, Polynomial1D):
+            coefficients = np.array(model.parameters)
+        elif isinstance(model, Polynomial2D):
+            degree = model.degree
+            coefficients = np.zeros((degree + 1, degree + 1))
+            for i in range(degree + 1):
+                for j in range(degree + 1):
+                    if i + j < degree + 1:
+                        name = "c" + str(i) + "_" + str(j)
+                        coefficients[i, j] = getattr(model, name).value
+
+        node = {"coefficients": coefficients}
+
+        if parse_tag_version(tag) >= self._DOMAIN_WINDOW_MIN_VERSION:
+            if model.n_inputs == 1:
+                if model.domain is not None:
+                    node["domain"] = model.domain
+                if model.window is not None:
+                    node["window"] = model.window
+            else:
+                if model.x_domain or model.y_domain is not None:
+                    node["domain"] = (model.x_domain, model.y_domain)
+                if model.x_window or model.y_window is not None:
+                    node["window"] = (model.x_window, model.y_window)
+
+        return node
+
+    def from_yaml_tree_transform(self, node, tag, ctx):
+        from astropy.modeling.polynomial import Polynomial1D, Polynomial2D
+
+        coefficients = np.asarray(node["coefficients"])
+        n_dim = coefficients.ndim
+
+        if n_dim == 1:
+            domain = node.get("domain", None)
+            window = node.get("window", None)
+
+            model = Polynomial1D(coefficients.size - 1,
+                                domain=domain, window=window)
+            model.parameters = coefficients
+        elif n_dim == 2:
+            x_domain, y_domain = tuple(node.get("domain", (None, None)))
+            x_window, y_window = tuple(node.get("window", (None, None)))
+            shape = coefficients.shape
+            degree = shape[0] - 1
+            if shape[0] != shape[1]:
+                raise TypeError("Coefficients must be an (n+1, n+1) matrix") # pragma: no cover
```

Comment:
Why marking `no cover`?  And below?

Reply:
Hmm.  Cheat to win?  This line is unreachable using only this library, I guess it was intended to check input from other implementations.  I'll remove the `no cover`.

Reply:
Yeah, there's a few more below.  😂 

I'm actually skeptical of 100% coverage projects, as it means the developers went through a lot of effort to get to that point, and probably had to do things like the above, or write bogus tests that make the coverage happy but don't actually test the code, i.e. when a test just reimplements the code it is supposed to be testing.  =)